File: src/engine/EventEngineImpl.ts
/**
 * @license
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  DeadLetterEntry,
  EngineEvent,
  EngineMetrics,
  EventHandler,
  EventMetadata,
  EventPriority,
  IEventEngine,
  Middleware,
  Subscription,
} from './types';

export class EventEngineImpl implements IEventEngine {
  private subscriptions: Map<string, Subscription> = new Map();
  private middlewares: Middleware[] = [];
  private history: EngineEvent<any>[] = [];
  private deadLetterQueue: Map<string, DeadLetterEntry> = new Map();
  private maxHistorySize: number;
  private startTime: number = Date.now();

  // Metrics state
  private metrics: EngineMetrics = {
    eventsPublished: 0,
    eventsProcessed: 0,
    eventsFailed: 0,
    eventsInDLQ: 0,
    activeSubscriptions: 0,
    avgDispatchLatencyMs: 0,
    peakThroughputPerSec: 0,
    totalContextMutations: 0,
    activeContextScopes: 0,
    activeRules: 0,
    uptimeSeconds: 0,
  };

  private latencySamples: number[] = [];
  private throughputWindow: number[] = [];
  private listeners: Set<(event: EngineEvent<any>) => void> = new Set();

  constructor(options?: { maxHistorySize?: number }) {
    this.maxHistorySize = options?.maxHistorySize ?? 1000;
    this.setupDefaultMiddlewares();
  }

  private setupDefaultMiddlewares(): void {
    // Built-in Telemetry & Sanitization Middleware
    this.use({
      name: 'telemetry-sanitizer',
      beforePublish: (event) => {
        if (!event.id) {
          event.id = `evt_${Date.now()}_${Math.random().toString(36).substring(2, 9)}`;
        }
        if (!event.metadata.timestamp) {
          event.metadata.timestamp = Date.now();
        }
        if (!event.metadata.priority) {
          event.metadata.priority = 'NORMAL';
        }
        if (!event.metadata.traceId) {
          event.metadata.traceId = `trc_${Math.random().toString(36).substring(2, 10)}`;
        }
        return event;
      },
      afterDispatch: (_event, _sub, durationMs) => {
        this.recordLatency(durationMs);
      },
      onError: (event, error) => {
        console.warn(`[EventEngine] Error in handler for topic '${event.topic}':`, error.message);
      },
    });
  }

  /**
   * Subscribe to internal event bus notifications (useful for UI live monitors)
   */
  public onAnyEvent(listener: (event: EngineEvent) => void): () => void {
    this.listeners.add(listener);
    return () => {
      this.listeners.delete(listener);
    };
  }

  public async publish<T = Record<string, unknown>>(
    topic: string,
    payload: T,
    metadata?: Partial<EventMetadata>
  ): Promise<EngineEvent<T>> {
    let event: EngineEvent<T> = {
      id: `evt_${Date.now()}_${Math.random().toString(36).substring(2, 9)}`,
      topic,
      payload,
      metadata: {
        timestamp: Date.now(),
        source: 'engine.core',
        priority: 'NORMAL',
        retryCount: 0,
        maxRetries: 3,
        ...metadata,
      },
      status: 'PENDING',
    };

    // Execute beforePublish middlewares
    for (const mw of this.middlewares) {
      if (mw.beforePublish) {
        try {
          const result = await mw.beforePublish(event);
          if (result === null) {
            event.status = 'SKIPPED';
            this.pushHistory(event);
            return event;
          }
          event = result as EngineEvent<T>;
        } catch (err: any) {
          console.error(`Middleware ${mw.name} failed on beforePublish:`, err);
        }
      }
    }

    this.metrics.eventsPublished++;
    this.recordThroughput();
    event.status = 'DISPATCHED';

    const startTime = performance.now();

    // Find matching subscriptions
    const matchingSubs = this.findMatchingSubscriptions(event.topic);
    
    // Sort subscriptions by priority descending
    matchingSubs.sort((a, b) => (b.priority ?? 0) - (a.priority ?? 0));

    let dispatchSuccess = true;
    let lastError: Error | undefined;

    if (matchingSubs.length === 0) {
      // Dispatched with no subscribers
      event.status = 'PROCESSED';
      event.executionTimeMs = Math.round((performance.now() - startTime) * 100) / 100;
      this.pushHistory(event);
      this.notifyListeners(event);
      return event;
    }

    for (const sub of matchingSubs) {
      if (!sub.active) continue;

      // Filter check
      if (sub.filter && !sub.filter(event)) {
        continue;
      }

      // Check middleware beforeDispatch
      let shouldProceed = true;
      for (const mw of this.middlewares) {
        if (mw.beforeDispatch) {
          try {
            const allow = await mw.beforeDispatch(event, sub);
            if (!allow) {
              shouldProceed = false;
              break;
            }
          } catch (e) {
            console.error(`Middleware ${mw.name} failed beforeDispatch`, e);
          }
        }
      }

      if (!shouldProceed) continue;

      const subStartTime = performance.now();
      try {
        await sub.handler(event);
        sub.invocationCount++;
        sub.lastInvokedAt = Date.now();

        const duration = performance.now() - subStartTime;
        for (const mw of this.middlewares) {
          if (mw.afterDispatch) {
            try {
              await mw.afterDispatch(event, sub, duration);
            } catch (e) {
              console.error(`Middleware ${mw.name} afterDispatch error:`, e);
            }
          }
        }
      } catch (err: any) {
        dispatchSuccess = false;
        lastError = err instanceof Error ? err : new Error(String(err));
        sub.errorCount++;

        for (const mw of this.middlewares) {
          if (mw.onError) {
            try {
              await mw.onError(event, lastError, sub);
            } catch (e) {
              console.error(`Middleware ${mw.name} onError failed:`, e);
            }
          }
        }
      }
    }

    const totalDuration = performance.now() - startTime;
    event.executionTimeMs = Math.round(totalDuration * 100) / 100;

    if (dispatchSuccess) {
      event.status = 'PROCESSED';
      this.metrics.eventsProcessed++;
    } else {
      event.status = 'FAILED';
      event.error = lastError?.message ?? 'Handler execution failed';
      this.metrics.eventsFailed++;
      this.handleDeadLetter(event, lastError?.message || 'Execution error');
    }

    this.pushHistory(event);
    this.notifyListeners(event);

    return event;
  }

  public async publishBatch(
    events: Array<{ topic: string; payload: any; metadata?: Partial<EventMetadata> }>
  ): Promise<EngineEvent[]> {
    // Sort batch by priority if provided
    const priorityWeight: Record<EventPriority, number> = {
      CRITICAL: 4,
      HIGH: 3,
      NORMAL: 2,
      LOW: 1,
    };

    const sortedEvents = [...events].sort((a, b) => {
      const pA = priorityWeight[a.metadata?.priority || 'NORMAL'];
      const pB = priorityWeight[b.metadata?.priority || 'NORMAL'];
      return pB - pA;
    });

    const results: EngineEvent[] = [];
    for (const item of sortedEvents) {
      const res = await this.publish(item.topic, item.payload, item.metadata);
      results.push(res);
    }
    return results;
  }

  public subscribe<T = Record<string, unknown>>(
    topicPattern: string,
    handler: EventHandler<T>,
    options?: {
      filter?: (event: EngineEvent<T>) => boolean;
      priority?: number;
      name?: string;
    }
  ): Subscription {
    const sub: Subscription = {
      id: `sub_${Date.now()}_${Math.random().toString(36).substring(2, 7)}`,
      topicPattern,
      handler,
      filter: options?.filter,
      priority: options?.priority ?? 0,
      name: options?.name ?? `Sub-${topicPattern}`,
      active: true,
      createdAt: Date.now(),
      invocationCount: 0,
      errorCount: 0,
    };

    this.subscriptions.set(sub.id, sub);
    this.metrics.activeSubscriptions = this.subscriptions.size;
    return sub;
  }

  public unsubscribe(subscriptionId: string): boolean {
    const deleted = this.subscriptions.delete(subscriptionId);
    this.metrics.activeSubscriptions = this.subscriptions.size;
    return deleted;
  }

  public use(middleware: Middleware): void {
    const existingIndex = this.middlewares.findIndex((m) => m.name === middleware.name);
    if (existingIndex >= 0) {
      this.middlewares[existingIndex] = middleware;
    } else {
      this.middlewares.push(middleware);
    }
  }

  public removeMiddleware(name: string): boolean {
    const index = this.middlewares.findIndex((m) => m.name === name);
    if (index >= 0) {
      this.middlewares.splice(index, 1);
      return true;
    }
    return false;
  }

  public getHistory(limit: number = 50, filterTopic?: string): EngineEvent[] {
    let list = [...this.history];
    if (filterTopic) {
      list = list.filter((e) => this.matchTopic(filterTopic, e.topic));
    }
    return list.slice(-limit);
  }

  public async replay(fromTimestampOrId: number | string, toTimestamp?: number): Promise<number> {
    let startIndex = -1;
    if (typeof fromTimestampOrId === 'string') {
      startIndex = this.history.findIndex((e) => e.id === fromTimestampOrId);
    } else {
      startIndex = this.history.findIndex((e) => e.metadata.timestamp >= fromTimestampOrId);
    }

    if (startIndex === -1) {
      return 0;
    }

    const eventsToReplay = this.history.slice(startIndex).filter((e) => {
      if (toTimestamp && e.metadata.timestamp > toTimestamp) {
        return false;
      }
      return true;
    });

    let replayedCount = 0;
    for (const evt of eventsToReplay) {
      await this.publish(evt.topic, evt.payload, {
        ...evt.metadata,
        correlationId: evt.id,
        source: `${evt.metadata.source || 'replayer'}:replay`,
      });
      replayedCount++;
    }

    return replayedCount;
  }

  public getDeadLetterQueue(): DeadLetterEntry[] {
    return Array.from(this.deadLetterQueue.values());
  }

  public async retryDeadLetter(dlqEntryId: string): Promise<boolean> {
    const entry = this.deadLetterQueue.get(dlqEntryId);
    if (!entry) return false;

    entry.retriesAttempted++;
    const maxRetries = entry.event.metadata.maxRetries ?? 3;

    try {
      const res = await this.publish(entry.event.topic, entry.event.payload, {
        ...entry.event.metadata,
        retryCount: entry.retriesAttempted,
      });

      if (res.status === 'PROCESSED') {
        this.deadLetterQueue.delete(dlqEntryId);
        this.metrics.eventsInDLQ = this.deadLetterQueue.size;
        return true;
      }
    } catch {
      // Continue tracking in DLQ
    }

    if (entry.retriesAttempted >= maxRetries) {
      entry.reason = `Max retries (${maxRetries}) exceeded: ${entry.reason}`;
    }

    this.metrics.eventsInDLQ = this.deadLetterQueue.size;
    return false;
  }

  public async retryAllDeadLetters(): Promise<number> {
    let success = 0;
    const entries = Array.from(this.deadLetterQueue.values());
    for (const entry of entries) {
      const ok = await this.retryDeadLetter(entry.id);
      if (ok) success++;
    }
    return success;
  }

  public clearDeadLetterQueue(): void {
    this.deadLetterQueue.clear();
    this.metrics.eventsInDLQ = 0;
  }

  public clearHistory(): void {
    this.history = [];
  }

  public getSubscriptions(): Subscription[] {
    return Array.from(this.subscriptions.values());
  }

  public getMetrics(): EngineMetrics {
    this.metrics.uptimeSeconds = Math.floor((Date.now() - this.startTime) / 1000);
    this.metrics.activeSubscriptions = this.subscriptions.size;
    this.metrics.eventsInDLQ = this.deadLetterQueue.size;
    return { ...this.metrics };
  }

  public resetMetrics(): void {
    this.metrics = {
      eventsPublished: 0,
      eventsProcessed: 0,
      eventsFailed: 0,
      eventsInDLQ: this.deadLetterQueue.size,
      activeSubscriptions: this.subscriptions.size,
      avgDispatchLatencyMs: 0,
      peakThroughputPerSec: 0,
      totalContextMutations: this.metrics.totalContextMutations,
      activeContextScopes: this.metrics.activeContextScopes,
      activeRules: this.metrics.activeRules,
      uptimeSeconds: 0,
    };
    this.latencySamples = [];
    this.throughputWindow = [];
    this.startTime = Date.now();
  }

  // --------------------------------------------------------------------------
  // Private Helpers
  // --------------------------------------------------------------------------

  private findMatchingSubscriptions(topic: string): Subscription[] {
    const matched: Subscription[] = [];
    for (const sub of this.subscriptions.values()) {
      if (this.matchTopic(sub.topicPattern, topic)) {
        matched.push(sub);
      }
    }
    return matched;
  }

  /**
   * Topic matching algorithm supports:
   * - Exact: "user.created"
   * - Single-segment wildcard: "order.*" matches "order.placed" but not "order.placed.vip"
   * - Multi-segment wildcard: "system.**" or "system.#" matches "system.auth.login.failed"
   * - Global wildcard: "*" or "**"
   */
  public matchTopic(pattern: string, topic: string): boolean {
    if (pattern === '*' || pattern === '**' || pattern === '#') return true;
    if (pattern === topic) return true;

    const patternParts = pattern.split('.');
    const topicParts = topic.split('.');

    let pIdx = 0;
    let tIdx = 0;

    while (pIdx < patternParts.length && tIdx < topicParts.length) {
      const pPart = patternParts[pIdx];

      if (pPart === '**' || pPart === '#') {
        // Multi-level wildcard matches rest of topic
        if (pIdx === patternParts.length - 1) return true;
        // Search next segment
        const nextP = patternParts[pIdx + 1];
        while (tIdx < topicParts.length && topicParts[tIdx] !== nextP) {
          tIdx++;
        }
        pIdx++;
        continue;
      }

      if (pPart === '*' || pPart === topicParts[tIdx]) {
        pIdx++;
        tIdx++;
        continue;
      }

      return false;
    }

    if (pIdx < patternParts.length && (patternParts[pIdx] === '**' || patternParts[pIdx] === '#')) {
      return true;
    }

    return pIdx === patternParts.length && tIdx === topicParts.length;
  }

  private handleDeadLetter(event: EngineEvent<any>, reason: string): void {
    const entryId = `dlq_${Date.now()}_${Math.random().toString(36).substring(2, 7)}`;
    this.deadLetterQueue.set(entryId, {
      id: entryId,
      event,
      failedAt: Date.now(),
      reason,
      retriesAttempted: event.metadata.retryCount || 0,
      stackTrace: event.error,
    });
    this.metrics.eventsInDLQ = this.deadLetterQueue.size;
  }

  private pushHistory(event: EngineEvent<any>): void {
    this.history.push(event);
    if (this.history.length > this.maxHistorySize) {
      this.history.shift();
    }
  }

  private recordLatency(ms: number): void {
    this.latencySamples.push(ms);
    if (this.latencySamples.length > 100) {
      this.latencySamples.shift();
    }
    const sum = this.latencySamples.reduce((a, b) => a + b, 0);
    this.metrics.avgDispatchLatencyMs = Math.round((sum / this.latencySamples.length) * 100) / 100;
  }

  private recordThroughput(): void {
    const now = Date.now();
    this.throughputWindow.push(now);
    // Keep window within last 1000ms
    this.throughputWindow = this.throughputWindow.filter((t) => now - t <= 1000);
    const count = this.throughputWindow.length;
    if (count > this.metrics.peakThroughputPerSec) {
      this.metrics.peakThroughputPerSec = count;
    }
  }

  private notifyListeners(event: EngineEvent<any>): void {
    for (const listener of this.listeners) {
      try {
        listener(event);
      } catch (err) {
        console.error('Error in event listener:', err);
      }
    }
  }
}

File: android-clean-architecture/src/main/kotlin/com/uru/data/autonomy/EventEngineImpl.kt
package com.autonomy.engine.data

import com.autonomy.engine.domain.*
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.LinkedBlockingQueue
import java.util.regex.Pattern

/**
 * High-performance, coroutine-powered EventEngine implementation for Android.
 */
class EventEngineImpl(
    private val scope: CoroutineScope = CoroutineScope(Dispatchers.Default + SupervisorJob())
) : EventEngine {

    private val subscriptions = ConcurrentHashMap<String, Subscription>()
    private val middlewares = CopyOnWriteArrayList<Middleware>()
    private val eventHistory = CopyOnWriteArrayList<EngineEvent<*>>()
    private val deadLetterQueue = ConcurrentHashMap<String, DeadLetterEntry>()
    private val maxHistorySize = 500

    private val _metrics = MutableStateFlow(EngineMetrics())
    override val metricsFlow: StateFlow<EngineMetrics> = _metrics.asStateFlow()

    private var totalLatencySumMs: Double = 0.0
    private var totalDispatchedEvents: Long = 0

    override suspend fun <T> publish(
        topic: String,
        payload: T,
        metadata: EventMetadata
    ): EngineEvent<T> {
        val startNs = System.nanoTime()

        var currentEvent: EngineEvent<*> = EngineEvent(
            topic = topic,
            payload = payload,
            metadata = metadata,
            status = EventStatus.PENDING
        )

        // Run beforePublish middlewares
        for (mw in middlewares) {
            try {
                val modified = mw.beforePublish(currentEvent)
                if (modified == null) {
                    val skipped = (currentEvent as EngineEvent<T>).copy(status = EventStatus.SKIPPED)
                    recordEvent(skipped)
                    return skipped
                }
                currentEvent = modified
            } catch (e: Exception) {
                mw.onError(currentEvent, e, null)
            }
        }

        // Find matching subscriptions sorted by priority (higher priority first)
        val matchingSubs = subscriptions.values
            .filter { it.active && matchTopic(it.topicPattern, currentEvent.topic) }
            .sortedByDescending { it.priority }

        if (matchingSubs.isEmpty()) {
            val processedEvent = (currentEvent as EngineEvent<T>).copy(
                status = EventStatus.PROCESSED,
                executionTimeMs = (System.nanoTime() - startNs) / 1_000_000
            )
            recordEvent(processedEvent)
            updateMetrics(published = 1, processed = 1, latencyMs = processedEvent.executionTimeMs?.toDouble() ?: 0.0)
            return processedEvent
        }

        var anyFailed = false
        var lastError: Throwable? = null

        for (sub in matchingSubs) {
            val subStart = System.currentTimeMillis()
            try {
                // Check filter if provided
                if (sub.filter != null && !sub.filter.invoke(currentEvent)) {
                    continue
                }

                // Middleware beforeDispatch
                var shouldDispatch = true
                for (mw in middlewares) {
                    if (!mw.beforeDispatch(currentEvent, sub)) {
                        shouldDispatch = false
                        break
                    }
                }
                if (!shouldDispatch) continue

                // Execute handler asynchronously
                sub.handler(currentEvent, null)
                sub.invocationCount++
                sub.lastInvokedAt = System.currentTimeMillis()

                val duration = System.currentTimeMillis() - subStart
                for (mw in middlewares) {
                    mw.afterDispatch(currentEvent, sub, duration)
                }
            } catch (e: Throwable) {
                anyFailed = true
                lastError = e
                sub.errorCount++
                for (mw in middlewares) {
                    mw.onError(currentEvent, e, sub)
                }
            }
        }

        val elapsedMs = (System.nanoTime() - startNs) / 1_000_000

        val finalEvent = if (anyFailed) {
            val failedEvent = (currentEvent as EngineEvent<T>).copy(
                status = EventStatus.FAILED,
                error = lastError?.message ?: "Unknown handler error",
                executionTimeMs = elapsedMs
            )
            moveToDeadLetter(failedEvent, lastError)
            updateMetrics(published = 1, failed = 1, latencyMs = elapsedMs.toDouble())
            failedEvent
        } else {
            val successfulEvent = (currentEvent as EngineEvent<T>).copy(
                status = EventStatus.PROCESSED,
                executionTimeMs = elapsedMs
            )
            updateMetrics(published = 1, processed = 1, latencyMs = elapsedMs.toDouble())
            successfulEvent
        }

        recordEvent(finalEvent)
        return finalEvent
    }

    override suspend fun publishBatch(
        events: List<Triple<String, Any?, EventMetadata>>
    ): List<EngineEvent<*>> {
        // Sort by priority order
        val priorityComparator = compareBy<Triple<String, Any?, EventMetadata>> { triple ->
            when (triple.third.priority) {
                EventPriority.CRITICAL -> 0
                EventPriority.HIGH -> 1
                EventPriority.NORMAL -> 2
                EventPriority.LOW -> 3
            }
        }

        val sorted = events.sortedWith(priorityComparator)
        val results = mutableListOf<EngineEvent<*>>()
        for (item in sorted) {
            results.add(publish(item.first, item.second, item.third))
        }
        return results
    }

    @Suppress("UNCHECKED_CAST")
    override fun <T> subscribe(
        topicPattern: String,
        options: SubscriptionOptions,
        handler: EventHandler<T>
    ): Subscription {
        val sub = Subscription(
            topicPattern = topicPattern,
            priority = options.priority,
            name = options.name,
            filter = options.filter,
            handler = { event, ctx ->
                handler(event as EngineEvent<T>, ctx)
            }
        )
        subscriptions[sub.id] = sub
        _metrics.value = _metrics.value.copy(activeSubscriptions = subscriptions.size)
        return sub
    }

    override fun unsubscribe(subscriptionId: String): Boolean {
        val removed = subscriptions.remove(subscriptionId) != null
        if (removed) {
            _metrics.value = _metrics.value.copy(activeSubscriptions = subscriptions.size)
        }
        return removed
    }

    override fun use(middleware: Middleware) {
        middlewares.add(middleware)
    }

    override fun removeMiddleware(name: String): Boolean {
        return middlewares.removeIf { it.name == name }
    }

    override fun getHistory(limit: Int, filterTopic: String?): List<EngineEvent<*>> {
        val list = if (filterTopic.isNullOrBlank()) {
            eventHistory.toList()
        } else {
            eventHistory.filter { matchTopic(filterTopic, it.topic) }
        }
        return list.takeLast(limit)
    }

    override suspend fun replay(fromTimestampOrId: Any, toTimestamp: Long?): Int {
        val eventsToReplay = eventHistory.filter { event ->
            when (fromTimestampOrId) {
                is Long -> event.metadata.timestamp >= fromTimestampOrId && (toTimestamp == null || event.metadata.timestamp <= toTimestamp)
                is String -> event.id == fromTimestampOrId || event.metadata.correlationId == fromTimestampOrId
                else -> false
            }
        }

        for (evt in eventsToReplay) {
            publish(
                topic = evt.topic,
                payload = evt.payload,
                metadata = evt.metadata.copy(
                    timestamp = System.currentTimeMillis(),
                    tags = evt.metadata.tags + "replayed"
                )
            )
        }
        return eventsToReplay.size
    }

    override fun getDeadLetterQueue(): List<DeadLetterEntry> {
        return deadLetterQueue.values.toList()
    }

    override suspend fun retryDeadLetter(dlqEntryId: String): Boolean {
        val entry = deadLetterQueue[dlqEntryId] ?: return false
        deadLetterQueue.remove(dlqEntryId)
        _metrics.value = _metrics.value.copy(eventsInDLQ = deadLetterQueue.size)

        val retried = publish(
            topic = entry.event.topic,
            payload = entry.event.payload,
            metadata = entry.event.metadata.copy(
                retryCount = entry.event.metadata.retryCount + 1,
                tags = entry.event.metadata.tags + "dlq_retry"
            )
        )
        return retried.status == EventStatus.PROCESSED
    }

    override suspend fun retryAllDeadLetters(): Int {
        val entries = deadLetterQueue.values.toList()
        var recovered = 0
        for (entry in entries) {
            if (retryDeadLetter(entry.id)) {
                recovered++
            }
        }
        return recovered
    }

    override fun clearDeadLetterQueue() {
        deadLetterQueue.clear()
        _metrics.value = _metrics.value.copy(eventsInDLQ = 0)
    }

    override fun clearHistory() {
        eventHistory.clear()
    }

    override fun getMetrics(): EngineMetrics = _metrics.value

    override fun resetMetrics() {
        totalLatencySumMs = 0.0
        totalDispatchedEvents = 0
        _metrics.value = EngineMetrics(activeSubscriptions = subscriptions.size)
    }

    private fun recordEvent(event: EngineEvent<*>) {
        eventHistory.add(event)
        if (eventHistory.size > maxHistorySize) {
            eventHistory.removeAt(0)
        }
    }

    private fun moveToDeadLetter(event: EngineEvent<*>, error: Throwable?) {
        val entry = DeadLetterEntry(
            event = event,
            reason = error?.message ?: "Execution failed",
            stackTrace = error?.stackTraceToString()
        )
        deadLetterQueue[entry.id] = entry
        _metrics.value = _metrics.value.copy(eventsInDLQ = deadLetterQueue.size)
    }

    private fun updateMetrics(published: Long = 0, processed: Long = 0, failed: Long = 0, latencyMs: Double = 0.0) {
        val current = _metrics.value
        totalDispatchedEvents += (processed + failed)
        totalLatencySumMs += latencyMs
        val avgLatency = if (totalDispatchedEvents > 0) totalLatencySumMs / totalDispatchedEvents else 0.0

        _metrics.value = current.copy(
            eventsPublished = current.eventsPublished + published,
            eventsProcessed = current.eventsProcessed + processed,
            eventsFailed = current.eventsFailed + failed,
            avgDispatchLatencyMs = (avgLatency * 100.0).toLong() / 100.0
        )
    }

    companion object {
        /**
         * Match topic string with wildcard support (*, **, #).
         * e.g., "sensor.*" matches "sensor.temp", "sensor.**" matches "sensor.room.1.temp".
         */
        fun matchTopic(pattern: String, topic: String): Boolean {
            if (pattern == topic || pattern == "#" || pattern == "**") return true

            val regexPattern = pattern
                .replace(".", "\\.")
                .replace("**", ".*")
                .replace("#", ".*")
                .replace("*", "[^.]+")

            return Pattern.compile("^$regexPattern$").matcher(topic).matches()
        }
    }
}
